File: 프로그래머스/Level 1/삼총사.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/131705
import Foundation

func solution(_ number:[Int]) -> Int {
    var answer:Int = 0
    var comArr:[[Int]] = []
    
    func combination(_ arr: [Int], _ cur: Int, _ depth: Int) {
        if depth == 3 {
            comArr.append(arr)
            return
        }
        
        for i in cur..<number.count {
            combination(arr+[number[i]], i+1, depth+1)
        }
    }
    
    combination([], 0, 0)
    
    for com in comArr {
        if com.reduce(0) {$0 + $1} == 0 {
            answer += 1
        }
    }
    return answer
}

File: 프로그래머스/Level 1/삼총사.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/131705
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int res = 0;
vector<int> visited;

void combination(vector<int> number, int tot, int cur, int depth) {
    if (depth == 3)
    {
        if (tot == 0) res++;
        return;
    }
    
    for (int i = cur; i < number.size(); ++i)
    {
        if (visited[i]) continue;
        visited[i] = 1;
        tot += number[i];
        combination(number, tot, i+1, depth+1);
        tot -= number[i];
        visited[i] = 0;
    }
    
    return;
}

int solution(vector<int> number) {
    int answer = 0;
    visited.resize(number.size(), 0);
    
    combination(number, 0, 0, 0);
    answer = res;
    
    return answer;
}